components/candle/progressionService.ts

```diff
@@ -0,0 +1,234 @@
+/*
+ *     The Peacock Project - a HITMAN server replacement.
+ *     Copyright (C) 2021-2023 The Peacock Project Team
+ *
+ *     This program is free software: you can redistribute it and/or modify
+ *     it under the terms of the GNU Affero General Public License as published by
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ *
+ *     This program is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU Affero General Public License for more details.
+ *
+ *     You should have received a copy of the GNU Affero General Public License
+ *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import { getSubLocationByName } from "../contracts/dataGen"
+import { controller } from "../controller"
+import { awardDropsToUser, getDataForUnlockables } from "../inventory"
+import type {
+    ContractSession,
+    UserProfile,
+    Unlockable,
+    GameVersion,
+} from "../types/types"
+import {
+    DEFAULT_MASTERY_MAXLEVEL,
+    clampValue,
+    xpRequiredForLevel,
+    xpRequiredForEvergreenLevel,
+    levelForXp,
+    evergreenLevelForXp,
+    getMaxProfileLevel,
+} from "../utils"
+import { writeUserData } from "../databaseHandler"
+
+export class ProgressionService {
+    grantProfileProgression(
+        actionXp: number,
+        masteryXp: number,
+        challengeDrops: Unlockable[],
+        contractSession: ContractSession,
+        userProfile: UserProfile,
+    ) {
+        // Grants profile XP
+        this.grantUserXp(actionXp + masteryXp, contractSession, userProfile)
+
+        // Grants Mastery Progression and Drops
+        this.grantLocationMasteryXpAndRewards(
+            masteryXp,
+            actionXp,
+            contractSession,
+            userProfile,
+        )
+
+        // Grants challenge related drops
+        awardDropsToUser(userProfile.Id, challengeDrops)
```

Comment:
`grantDrops` is better (it's literally in the comment). We're in the progressionService, it's pretty clear it's for the user.